pkg/chainreader/contract_reader_test.go

```diff
@@ -0,0 +1,308 @@
+package chainreader
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils/tests"
+)
+
+func TestContractReaderByIDsUnbind(t *testing.T) {
+	ctx := tests.Context(t)
+	mockReader := &ContractReaderMock{}
+	crByIDs := &ContractReaderByIDs{
+		bindings: sync.Map{},
+		cr:       mockReader,
+	}
+
+	bc1 := types.BoundContract{Address: "0x123", Name: "testContract1"}
+	bc2 := types.BoundContract{Address: "0x321", Name: "testContract2"}
+	bc1CustomID, bc2CustomID := "customID1", "customID2"
+
+	// Mock Bind function
+	mockReader.bindFunc = func(ctx context.Context, contracts []types.BoundContract) error {
+		expectedContracts := []types.BoundContract{bc1, bc2}
+		assert.ElementsMatch(t, expectedContracts, contracts)
+		return nil
+	}
+	require.NoError(t, crByIDs.Bind(ctx, map[string]types.BoundContract{bc1CustomID: bc1, bc2CustomID: bc2}))
+
+	// Mock Unbind function for error case
+	mockReader.unbindFunc = func(ctx context.Context, contracts []types.BoundContract) error {
+		return fmt.Errorf("some error")
+	}
+
+	// Test Unbind with error shouldn't remove bindings
+	require.Error(t, crByIDs.Unbind(ctx, map[string]types.BoundContract{bc1CustomID: bc1}))
```

Comment:
/nit is this slightly clearer for the reader if the map is empty?
```suggestion
	require.Error(t, crByIDs.Unbind(ctx, map[string]types.BoundContract{}))
```

Reply:
Ah, no, I think it's clearer as is given the Test successful Unbind case